confuse.py

```diff
@@ -134,36 +135,193 @@ def __init__(self, filename, reason=None):
 
 # Views and sources.
 
+
+UNSET = object()  # sentinel
+
+
+def _load_first(func):
+    '''Call self.load() before the function is called - used for lazy source
+    loading'''
+    def inner(self, *a, **kw):
+        self.load()
+        return func(self, *a, **kw)
+
+    try:
+        return wraps(func)(inner)
+    except AttributeError:
+        # in v2 they don't ignore missing attributes
+        # v3: https://github.com/python/cpython/blob/3.8/Lib/functools.py
+        # v2: https://github.com/python/cpython/blob/2.7/Lib/functools.py
+        inner.__name__ = func.__name__
+        return inner
+
+
+def all_subclasses(cls):
+    return set(cls.__subclasses__()).union(
+        s for c in cls.__subclasses__() for s in all_subclasses(c))
+
+
 class ConfigSource(dict):
-    """A dictionary augmented with metadata about the source of the
+    '''A dictionary augmented with metadata about the source of the
     configuration.
-    """
-    def __init__(self, value, filename=None, default=False):
-        super(ConfigSource, self).__init__(value)
+    '''
+    def __init__(self, value=UNSET, filename=None, default=False):
+        # track whether a config source has been set yet
+        self.loaded = value is not UNSET
+        super(ConfigSource, self).__init__(value if self.loaded else {})
         if filename is not None and not isinstance(filename, BASESTRING):
             raise TypeError(u'filename must be a string or None')
         self.filename = filename
         self.default = default
 
     def __repr__(self):
-        return 'ConfigSource({0!r}, {1!r}, {2!r})'.format(
-            super(ConfigSource, self),
-            self.filename,
-            self.default,
-        )
+        return '{}({}, filename={}, default={})'.format(
+            self.__class__.__name__,
+            dict.__repr__(self)
+            if self.loaded else '[Unloaded]'
+            if self.exists else "[Source doesn't exist]",
+            self.filename, self.default)
+
+    @property
+    def exists(self):
+        '''Does this config have access to usable configuration values?'''
+        return self.loaded or self.filename and os.path.isfile(self.filename)
+
+    def load(self):
+        '''Ensure that the source is loaded.'''
+        if not self.loaded:
+            self.config_dir()
+            self.loaded = self._load() is not False
```

Comment:
Similarly, I'm a tad confused about why this could ever be false… shouldn't `load` either always succeed or throw an exception? It seems like things could get pretty confusing if sources are allowed to silently fail.

Or if the idea is that missing files should be loadable and just silently produce no values (i.e., an empty dict), that's cool too… but perhaps that should be optional? (And it should still mark the source as "loaded" so that we don't keep trying?)

Reply:
That was a mechanism to allow a source subclass to signify that a source wasn't loaded and to try again next time. It's only if a `_load` function explicitly returns False. We can take this out if it's not desired.

Whether they are silently ignored is optional and disabled by default (`YamlSource(path, ignore_missing=True)` will ignore a source if it's missing). I only added it in because that's how the user config was handled.